### PR TITLE
Rethink STM32 I2C v2 HAL

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

This PR is a major rethink of the STM32 I2C code for their v2 I2C peripheral.  It tries to get it into a more API-compliant and less buggy state.  It should fix several known bugs:
- Single-byte API never returns NACK result code on error, only timeout (#61)
- Single-byte API returns errors one byte later than the byte that actually caused the error (#61)
- Using the single-byte API puts the I2C peripheral in a bad state where the next operation will fail (#62)

Additionally, I found more issues while making this PR that will also be fixed:
- Doing a single-byte operation that fails, then doing a transaction operation after that causes the transaction to return early and blindly report success, because the STOPF flag from earlier was not cleared 
- Doing an asynchronous read transaction then a synchronous write transaction could cause the asynchronous read transaction to be done a second time due to incorrect logic in the interrupt handler
- The entire I2C peripheral was being reinitialized whenever a NACK was received, which is a fairly expensive operation, especially for an interrupt handler.  Now NACKs are handled differently from other errors and don't cause a complete re-init.
- Doing a repeated start with the single-byte API did not work, because the RELOAD and NBYTES fields have to be zeroed before the repeated start can be sent.

The biggest change this PR makes is the addition of a new state enum, `obj_s->state`.  This provides a simple indicator of what the I2C code is doing at any given time, making certain parts of the logic a lot simpler and enabling some of the fixes to work.  I also added a lot of new checks based on this state variable, so if you use the I2C object in an incorrect way, you have a fighting chance of getting a useful error code back instead of undefined behavior.  (it's still tough though because certain functions do not have the ability to return an error code...)

#### Impact of changes 
Lots of issues are fixed (see above)

### Documentation 

None

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR

Working on creating new tests that run this code against hardware.
    
----------------------------------------------------------------------------------------------------------------
### Reviewers 

----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: targets/TARGET_STM/TARGET_STM32F0/objects.h [hunk 1]
```diff
@@ -77,39 +77,6 @@ struct serial_s {
 #endif
 };
 
-struct i2c_s {
```
Comment (multiplemonomials) [inline targets/TARGET_STM/TARGET_STM32F0/objects.h:80]: These were all consolidated into a single definition in stm_i2c_api.h.

File: features/frameworks/unity/source/unity.c [hunk 2]
```diff
@@ -17,7 +17,7 @@ int UNITY_OUTPUT_FLUSH(void);
 #endif
 
 /* Helpful macros for us to use here */
-#define UNITY_FAIL_AND_BAIL   { UNITY_OUTPUT_CHAR('\n'); utest_unity_assert_failure(); }
+#define UNITY_FAIL_AND_BAIL   { UNITY_OUTPUT_CHAR('\n'); utest_unity_assert_failure(); Unity.CurrentTestFailed = 1; }
```
Comment (multiplemonomials) [inline features/frameworks/unity/source/unity.c:20]: This fixes the test suite printing PASS after tests that fail.  No idea why this was not being set before.

File: targets/TARGET_STM/i2c_api.c [hunk 3]
```diff
@@ -1009,8 +1056,18 @@ int i2c_byte_write(i2c_t *obj, int data)
...
+    if(!(obj_s->state == STM_I2C_SB_WRITE_IN_PROGRESS || obj_s->state == STM_I2C_PENDING_START))
+    {
+        // Must either be ready to send address or ready to write data bytes
+        return 3;
+    }
+
+    if (obj_s->state == STM_I2C_PENDING_START)
+    {
+        // Clear Acknowledge Failure flag in case it was set from a previous operation
+        __HAL_I2C_CLEAR_FLAG(handle, I2C_FLAG_AF);
```
Comment (multiplemonomials) [inline targets/TARGET_STM/i2c_api.c:1069]: TODO I think this should be removed, but it makes tests fail when I do...

**Comment by multiplemonomials** [comment]
I have created an exhaustive new test suite that covers just about every pattern of I2C reads and writes you could do, including repeated starts and use of all three different APIs on the same peripheral one after the other.  It helped me catch a large number of issues in the existing and the new code.

https://github.com/multiplemonomials/Mbed-CI-Test-Shield/blob/master/Software/I2CBasicTest.cpp

**Comment by multiplemonomials** [comment]
Tests passed on STM32L452RE:
```
1/1 Test #4: testshield-i2c-basic .............   Passed   11.40 sec

The following tests passed:
        testshield-i2c-basic

100% tests passed, 0 tests failed out of 1
```

Tests passed on STM32H743ZI:
```
1/1 Test #4: testshield-i2c-basic .............   Passed    7.42 sec

The following tests passed:
        testshield-i2c-basic

100% tests passed, 0 tests failed out of 1
```

**Comment by multiplemonomials** [comment]
Almost there, just need to look into why this PR makes the "Simple Read - Transaction" test fail for IP v1

**Comment by multiplemonomials** [comment]
OK, I can now say that there are no more failures with IP v1 on this branch than on master.  Filed #96 about the remaining test failures.

File: targets/TARGET_STM/TARGET_STM32F0/objects.h (hunk 1 above)
Comment (JohnK1987) [reply targets/TARGET_STM/TARGET_STM32F0/objects.h:80]: Wow, multiple times same, interesting. Good job.

**Review by JohnK1987 — DISMISSED** [review]

**Comment by JojoS62** [comment]
it is difficult to test, I can connect some I2C devices and test with my currenct hardware. Maybe there is some I2C compliance test equiment, but I don't have access to it.

**Comment by JohnK1987** [comment]
@JojoS62, I believe it is about this piece of hardware - [Mbed-CI-Test-Shield](https://github.com/multiplemonomials/Mbed-CI-Test-Shield) by @multiplemonomials.

**Comment by multiplemonomials** [comment]
Yup!  I can send you one of the boards if you want to run these tests locally

**Comment by JohnK1987** [comment]
I would like also one, but I am from Czech republic and the transport from USA is probably expensive.
However, my employer has a PCB milling machine, I have it in my office. But for that is necessary have source files in Autodesk Eagle.

BR, Jan

**Comment by JojoS62** [comment]
and I'm living in germany. So it is probably easier to run some simple tests with I2C devices. Although, target implementations can have very different behaviour. I have seen that with SPI on H7, it can be very complicated.

File: targets/TARGET_STM/i2c_api.c (hunk 3 above)
Comment (multiplemonomials) [reply targets/TARGET_STM/i2c_api.c]: Removed this, test still pass.  I think the issue was something else, one of the race conditions found later causing random failures.

**Comment by JohnK1987** [comment]
I saw your work was already Marged also in ArmMbed. May I close issues (61,62 and 94 mentioned above) connected to this?

BR, Jan

**Comment by multiplemonomials** [comment]
Sure on 61 and 62.  I haven't fixed 94 yet.

**Pull request merged**